src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts

```diff
@@ -0,0 +1,231 @@
+import { EmailService } from '@seedcompany/nestjs-email';
+import { uniqBy } from 'lodash';
+import { ID, Role, Session, UnsecuredDto } from '~/common';
+import {
+  ConfigService,
+  EventsHandler,
+  IEventHandler,
+  ILogger,
+  Logger,
+} from '~/core';
+import {
+  EmailReportStatusNotification,
+  ProgressReportStatusChanged,
+} from '~/core/email/templates/progress-report-status-changed.template';
+import { AuthenticationService } from '../../../authentication';
+import { LanguageService } from '../../../language';
+import { PeriodicReportService } from '../../../periodic-report';
+import { ProjectService } from '../../../project';
+import { UserService } from '../../../user';
+import { ProgressReportStatus as Status } from '../../dto';
+import { ProgressReportWorkflowEvent } from '../dto/workflow-event.dto';
+import { WorkflowUpdatedEvent } from '../events/workflow-updated.event';
+import { ProgressReportWorkflowRepository } from '../progress-report-workflow.repository';
+import { ProgressReportWorkflowService } from '../progress-report-workflow.service';
+import { InternalTransition } from '../transitions';
+
+interface Member {
+  id?: ID | undefined;
+  email?: string | undefined;
+  role?: Role[] | undefined;
+}
+type Recipient = Omit<Member, 'role'>;
+
+@EventsHandler(WorkflowUpdatedEvent)
+export class ProgressReportWorkflowNotificationHandler
+  implements IEventHandler<WorkflowUpdatedEvent>
+{
+  constructor(
+    private readonly auth: AuthenticationService,
+    private readonly repo: ProgressReportWorkflowRepository,
+    private readonly configService: ConfigService,
+    private readonly userService: UserService,
+    private readonly projectService: ProjectService,
+    private readonly languageService: LanguageService,
+    private readonly reportService: PeriodicReportService,
+    private readonly emailService: EmailService,
+    private readonly workflowService: ProgressReportWorkflowService,
+    @Logger('ProgressReportWorkflowNotificationHandler:report-status-changed')
+    private readonly logger: ILogger,
+  ) {}
+
+  async handle({
+    reportId,
+    previousStatus,
+    next,
+    session,
+    workflowEvent,
+  }: WorkflowUpdatedEvent) {
+    if (!this.configService.sendProgressReportNotifications) {
+      return;
+    }
+
+    const membersWithValidRoles = await this.getTeamMembersWithValidRoles(
+      reportId,
+      next,
+    );
+
+    const envEmailList = this.getEnvList(next) as [];
+    const usersInEnvList: Recipient[] = await Promise.all(
+      envEmailList.map(async (item) => {
+        const user = await this.repo.getUserInfoByEmail(item);
```

Comment:
It would be better grab all the users from email addresses at once.
```ts
const users = await this.repo.getUserIdByEmails(envEmailList);
const userIdByEmail: Record<string, ID> = mapFromList([
  ...envEmailList.map(email => ({ id: undefined, email }), // defaults for missing users
  ...users,
], (({ id, email }) => [email, id]);
```

Reply:
I'm having trouble figuring out the new Neo4j cypher query for the new query by array that you prefer to use.  I've tried several different ways but none have worked so far.  This is the latest that I've tried so far, but still no success; please let me know what the correct query should be:
 `  async getUserIdByEmails(emails: string[]) {
    const query = this.db
      .query()
      .match([
        node('email', 'EmailAddress'),
        relation('in', '', 'email', ACTIVE),
        node('user', 'User'),
      ])
      .where({ 'user.email': inArray(emails) })
      .return<{
        users?: User[];
      }>(['user as users']);
    return await query.run();
  }`

Reply:
That's close!
```diff
- .where({ 'user.email': inArray(emails) })
+ .where({ 'email.value': inArray(emails) })
```
Then
```ts
.return<{ id: ID, email: string }>([
	'user.id as id',
	'email.value as email',
])
```